gossipsub.go

```diff
@@ -208,21 +230,29 @@ func (gs *GossipSubRouter) handlePrune(p peer.ID, ctl *pb.ControlMessage) {
 	}
 }
 
-func (gs *GossipSubRouter) Publish(from peer.ID, msg *pb.Message) {
-	gs.mcache.Put(msg)
-
-	tosend := make(map[peer.ID]struct{})
+func (gs *GossipConfigurableRouter) Publish(from peer.ID, msg *pb.Message) {
 	for _, topic := range msg.GetTopicIDs() {
+		config, ok := gs.topicConfigs[topic]
+		if ok {
+			config.Publish(gs, from, msg)
+		}
+	}
+}
+
+func (gs *GossipConfigurableRouter) AddGossipPeers(tosend map[peer.ID]struct{}, topics []string, withFloodSubPeers bool){
+	for _, topic := range topics {
 		// any peers in the topic?
 		tmap, ok := gs.p.topics[topic]
 		if !ok {
 			continue
 		}
 
-		// floodsub peers
-		for p := range tmap {
-			if gs.peers[p] == FloodSubID {
-				tosend[p] = struct{}{}
+		if withFloodSubPeers{
```

Comment:
The various configurations that I've implemented so far could likely all be done in floodsub, episub, etc.  Two caveats are 1) we'd need to allow the layer above the router implementations to have more access to the internals (e.g. which peers I'm syncing/sending messages to for a given topic, some control over limiting or manually triggering a broadcast) 2) The existing less performant implementations (e.g. floodsub) would do extremely poorly for some of these configuration options (e.g. persistence) anyway.

Additionally, in future implementations of say episub the configs used by gossipsub should be reusable.

Reply:
Given that we're ok breaking the router implementations/interfaces a bit do you think I should try and setup some of the configurations (e.g. persistence) to work across all router implementations or is leaving the configurations as something to be plugged into future routers (e.g. episub) sufficient?

Reply:
that's fine, although I don't know how useful it will be for floodsub.